File: app/src/main/java/com/ultron/assistant/accessibility/UltronAccessibilityService.java
package com.ultron.assistant.accessibility;

import android.accessibilityservice.AccessibilityService;
import android.accessibilityservice.GestureDescription;
import android.graphics.Path;
import android.os.Bundle;
import android.view.accessibility.AccessibilityNodeInfo;

import java.util.List;

public class UltronAccessibilityService
        extends AccessibilityService {private static UltronAccessibilityService instance;

    @Override
    public void onServiceConnected() {
        super.onServiceConnected();
        instance = this;
    }

    @Override
    public void onAccessibilityEvent(
            android.view.accessibility.AccessibilityEvent event
    ) {
        // Service monitors active windows.
    }

    @Override
    public void onInterrupt() {
    }

    @Override
    public void onDestroy() {
        instance = null;
        super.onDestroy();
    }

    public static boolean isRunning() {
        return instance != null;
    }

    public static boolean clickText(String text) {

        if (instance == null || text == null || text.trim().isEmpty()) {
            return false;
        }

        String target = text.trim().toLowerCase();

        AccessibilityNodeInfo root =
                instance.getRootInActiveWindow();

        if (root == null) {
            return false;
        }

        try {
            // First try Android's built-in text search
            List<AccessibilityNodeInfo> nodes =
                    root.findAccessibilityNodeInfosByText(text.trim());

            if (nodes != null) {
                for (AccessibilityNodeInfo node : nodes) {
                    if (clickNodeOrParent(node)) {
                        return true;
                    }
                }
            }

            // Fallback: search all visible nodes manually
            return findAndClickMatchingNode(root, target);

        } finally {
            root.recycle();
        }
    }

    private static boolean findAndClickMatchingNode(
            AccessibilityNodeInfo node,
            String target
    ) {

        if (node == null) {
            return false;
        }

        CharSequence nodeText = node.getText();
        CharSequence description = node.getContentDescription();

        if (matchesText(nodeText, target)
                || matchesText(description, target)) {

            if (clickNodeOrParent(node)) {
                return true;
            }
        }

        for (int i = 0; i < node.getChildCount(); i++) {

            AccessibilityNodeInfo child = node.getChild(i);

            if (child != null) {
                if (findAndClickMatchingNode(child, target)) {
                    return true;
                }
            }
        }

        return false;
    }

    private static boolean matchesText(
            CharSequence value,
            String target
    ) {

        if (value == null || target == null) {
            return false;
        }

        String current =
                value.toString().trim().toLowerCase();

        return current.equals(target)
                || current.contains(target)
                || target.contains(current);
    }

    private static boolean clickNodeOrParent(
            AccessibilityNodeInfo node
    ) {

        AccessibilityNodeInfo current = node;

        while (current != null) {

            if (current.isClickable()
                    && current.isEnabled()
                    && current.isVisibleToUser()) {

                return current.performAction(
                        AccessibilityNodeInfo.ACTION_CLICK
                );
            }

            current = current.getParent();
        }

        return false;
    }

    public static boolean typeText(String text) {

        if (instance == null || text == null) {
            return false;
        }

        AccessibilityNodeInfo root =
                instance.getRootInActiveWindow();

        if (root == null) {
            return false;
        }

        try {
            AccessibilityNodeInfo focused =
                    root.findFocus(
                            AccessibilityNodeInfo.FOCUS_INPUT
                    );

            if (focused == null) {
                return false;
            }

            Bundle arguments = new Bundle();

            arguments.putCharSequence(
                    AccessibilityNodeInfo
                            .ACTION_ARGUMENT_SET_TEXT_CHARSEQUENCE,
                    text
            );

            return focused.performAction(
                    AccessibilityNodeInfo.ACTION_SET_TEXT,
                    arguments
            );

        } finally {
            root.recycle();
        }
    }

    public static boolean clickSearchField() {

        if (instance == null) {
            return false;
        }

        AccessibilityNodeInfo root =
                instance.getRootInActiveWindow();

        if (root == null) {
            return false;
        }

        try {
            AccessibilityNodeInfo node =
                    findEditableNode(root);

            if (node != null) {

                boolean clicked =
                        node.performAction(
                                AccessibilityNodeInfo.ACTION_CLICK
                        );

                node.performAction(
                        AccessibilityNodeInfo.ACTION_FOCUS
                );

                return clicked;
            }

        } finally {
            root.recycle();
        }

        return false;
    }

    private static AccessibilityNodeInfo findEditableNode(
            AccessibilityNodeInfo node
    ) {

        if (node == null) {
            return null;
        }

        if (node.isEditable()) {
            return node;
        }

        for (int i = 0;
             i < node.getChildCount();
             i++) {

            AccessibilityNodeInfo child =
                    node.getChild(i);

            AccessibilityNodeInfo result =
                    findEditableNode(child);

            if (result != null) {
                return result;
            }
        }

        return null;
    }

    public static boolean goBack() {

        return instance != null
                && instance.performGlobalAction(
                        GLOBAL_ACTION_BACK
                );
    }

    public static boolean goHome() {

        return instance != null
                && instance.performGlobalAction(
                        GLOBAL_ACTION_HOME
                );
    }

    public static boolean scrollDown() {

        if (instance == null) {
            return false;
        }

        AccessibilityNodeInfo root =
                instance.getRootInActiveWindow();

        if (root == null) {
            return false;
        }

        try {
            return findAndScroll(
                    root,
                    AccessibilityNodeInfo.ACTION_SCROLL_FORWARD
            );

        } finally {
            root.recycle();
        }
    }

    public static boolean scrollUp() {

        if (instance == null) {
            return false;
        }

        AccessibilityNodeInfo root =
                instance.getRootInActiveWindow();

        if (root == null) {
            return false;
        }

        try {
            return findAndScroll(
                    root,
                    AccessibilityNodeInfo.ACTION_SCROLL_BACKWARD
            );

        } finally {
            root.recycle();
        }
    }

    private static boolean findAndScroll(
            AccessibilityNodeInfo node,
            int action
    ) {

        if (node == null) {
            return false;
        }

        if (node.isScrollable()) {
            return node.performAction(action);
        }

        for (int i = 0;
             i < node.getChildCount();
             i++) {

            AccessibilityNodeInfo child =
                    node.getChild(i);

            if (findAndScroll(child, action)) {
                return true;
            }
        }

        return false;
    }
}

File: app/src/main/java/com/ultron/assistant/accessibility/UltronAccessibilityService.kt
package com.ultron.assistant.accessibility

import android.accessibilityservice.AccessibilityService
import android.accessibilityservice.GestureDescription
import android.content.Intent
import android.graphics.Path
import android.os.Bundle
import android.view.accessibility.AccessibilityEvent
import android.view.accessibility.AccessibilityNodeInfo

class UltronAccessibilityService : AccessibilityService() {

    companion object {
        var instance: UltronAccessibilityService? = null

        fun get(): UltronAccessibilityService? = instance
    }

    override fun onServiceConnected() {
        super.onServiceConnected()
        instance = this
    }

    override fun onAccessibilityEvent(event: AccessibilityEvent?) {
        // Future hook
    }

    override fun onInterrupt() {
    }

    override fun onDestroy() {
        instance = null
        super.onDestroy()
    }

    // APP CONTROL
    fun openApp(packageName: String): Boolean {
        return try {
            val intent = packageManager.getLaunchIntentForPackage(packageName)
            if (intent != null) {
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                startActivity(intent)
                true
            } else {
                false
            }
        } catch (e: Exception) {
            false
        }
    }

    // SCREEN TAP
    fun tap(x: Float, y: Float): Boolean {
        val path = Path()
        path.moveTo(x, y)
        val gesture = GestureDescription.Builder()
            .addStroke(GestureDescription.StrokeDescription(path, 0, 80))
            .build()
        return dispatchGesture(gesture, null, null)
    }

    // SWIPE
    fun swipe(startX: Float, startY: Float, endX: Float, endY: Float, duration: Long = 400): Boolean {
        val path = Path()
        path.moveTo(startX, startY)
        path.lineTo(endX, endY)
        val gesture = GestureDescription.Builder()
            .addStroke(GestureDescription.StrokeDescription(path, 0, duration))
            .build()
        return dispatchGesture(gesture, null, null)
    }

    // SYSTEM NAVIGATION
    fun goBack(): Boolean = performGlobalAction(GLOBAL_ACTION_BACK)
    fun goHome(): Boolean = performGlobalAction(GLOBAL_ACTION_HOME)
    fun openRecents(): Boolean = performGlobalAction(GLOBAL_ACTION_RECENTS)

    // YOUTUBE MEDIA CONTROLS
    fun youtubePlayPause(): Boolean = performGlobalAction(GLOBAL_ACTION_PLAY_PAUSE)
    fun youtubeNext(): Boolean = performGlobalAction(GLOBAL_ACTION_MEDIA_NEXT)
    fun youtubePrevious(): Boolean = performGlobalAction(GLOBAL_ACTION_MEDIA_PREVIOUS)

    // UI NODE CLICK
    fun clickText(text: String): Boolean {
        val root = rootInActiveWindow ?: return false
        val nodes = root.findAccessibilityNodeInfosByText(text)
        for (node in nodes) {
            if (node.isClickable) {
                return node.performAction(AccessibilityNodeInfo.ACTION_CLICK)
            }
            var parent = node.parent
            while (parent != null) {
                if (parent.isClickable) {
                    return parent.performAction(AccessibilityNodeInfo.ACTION_CLICK)
                }
                parent = parent.parent
            }
        }
        return false
    }

    // SET TEXT IN INPUT FIELD
    fun setText(text: String): Boolean {
        val root = rootInActiveWindow ?: return false
        val focusedNode = root.findFocus(AccessibilityNodeInfo.FOCUS_INPUT) ?: return false
        val arguments = Bundle()
        arguments.putCharSequence(
            AccessibilityNodeInfo.ACTION_ARGUMENT_SET_TEXT_CHARSEQUENCE,
            text
        )
        return focusedNode.performAction(AccessibilityNodeInfo.ACTION_SET_TEXT, arguments)
    }
}

